addons/account_edi/models/account_edi_format.py

```diff
@@ -512,33 +513,28 @@ def search_with_vat(extra_domain):
 
             # Sometimes, the vat is specified with some whitespaces.
             normalized_vat = vat.replace(' ', '')
-            country_prefix = None
+            country_prefix = re.match('^[a-zA-Z]{2}|^', vat).group()
 
-            partner = self.env['res.partner'].search(extra_domain + [('vat', '=', normalized_vat)], limit=1)
+            partner = self.env['res.partner'].search(extra_domain + [('vat', 'in', (normalized_vat, vat))], limit=1)
 
             # Try to remove the country code prefix from the vat.
-            if not partner and len(normalized_vat) > 2 and normalized_vat[:2].isalpha():
-                country_prefix = normalized_vat[:2]
-                normalized_vat = normalized_vat[2:]
+            if not partner and country_prefix:
                 partner = self.env['res.partner'].search(extra_domain + [
-                    ('vat', '=', normalized_vat),
-                    ('country_id.code', '=', country_prefix.lower()),
+                    ('vat', 'in', (normalized_vat[2:], vat[2:])),
+                    ('country_id.code', '=', country_prefix.upper()),
                 ], limit=1)
 
                 # The country could be not specified on the partner.
                 if not partner:
                     partner = self.env['res.partner'].search(extra_domain + [
-                        ('vat', '=', normalized_vat),
+                        ('vat', 'in', (normalized_vat[2:], vat[2:])),
                         ('country_id', '=', False),
                     ], limit=1)
 
             # The vat could be a string of alphanumeric values without country code but with missing zeros at the
             # beginning.
             if not partner:
-                try:
-                    vat_only_numeric = str(int(normalized_vat))
-                except ValueError:
-                    vat_only_numeric = None
+                vat_only_numeric = int(re.sub('\D', '', vat) or 0)
```

Comment:
Note that the regex search below is bypassing the index https://github.com/odoo/odoo/blob/d6eb4752d7fbd22dd9ad744db51444e1392fd1bb/odoo/addons/base/models/res_partner.py#L173

This might be really bad for big databases with millions of partners.